### PR TITLE
rotations: make optional, add function to query availability.

### PR DESCRIPTION
**Pull request by matz-e**
SONATA's rotations are optional and may be specified as angles. Account
for this, with small additions in the API.

### CODE REVIEW

File: include/mvd/bits/sonata_misc.hpp [hunk 1]
```diff
@@ -71,24 +73,68 @@ inline Positions SonataFile::getPositions(const Range & range) const{
...
-    auto ws = pop_->getAttribute<double>("orientation_w", select(range, size_));
+    Positions res{boost::extents[range.count > 0 ? range.count : size_][4]};
 
-    // No direct slicing write access from std::vector
-    for (size_t i = 0; i < (range.count > 0 ? range.count : size_); ++i) {
-        res[i][0] = xs[i];
-        res[i][1] = ys[i];
-        res[i][2] = zs[i];
-        res[i][3] = ws[i];
+    const auto attrs = pop_->attributeNames();
```
Comment (ferdonline) [inline include/mvd/bits/sonata_misc.hpp]: I would split this routine. At least quat = .... is done twice with the same code

File: include/mvd/bits/sonata_misc.hpp [hunk 2]
```diff
@@ -71,24 +73,68 @@ inline Positions SonataFile::getPositions(const Range & range) const{
...
+                       attrs.count("orientation_z") +
+                       attrs.count("orientation_w")) == 4;
+
+    if (quat) {
+        auto xs = pop_->getAttribute<double>("orientation_x", select(range, size_));
+        auto ys = pop_->getAttribute<double>("orientation_y", select(range, size_));
+        auto zs = pop_->getAttribute<double>("orientation_z", select(range, size_));
+        auto ws = pop_->getAttribute<double>("orientation_w", select(range, size_));
+
+        // No direct slicing write access from std::vector
```
Comment (ferdonline) [inline include/mvd/bits/sonata_misc.hpp]: Tried using boost::multi_array_ref to wrap the vector? 
Also, I saw some examples with array_view for assignment in https://theboostcpplibraries.com/boost.multiarray

File: include/mvd/bits/sonata_misc.hpp [hunk 3]
```diff
@@ -71,24 +73,68 @@ inline Positions SonataFile::getPositions(const Range & range) const{
...
+        for (size_t i = 0; i < (range.count > 0 ? range.count : size_); ++i) {
+            res[i][0] = xs[i];
+            res[i][1] = ys[i];
+            res[i][2] = zs[i];
+            res[i][3] = ws[i];
+        }
+    } else {
+        for (size_t i = 0; i < (range.count > 0 ? range.count : size_); ++i) {
+            Quaternion rot{1., 0., 0., 0.};
```
Comment (ferdonline) [inline include/mvd/bits/sonata_misc.hpp]: Nice. I would just promote this mathematical conversion part to its own routine

**Comment by ferdonline** [comment]
Noticed your name is not in the hall of (in)fame (README.md). It would be reasonable to be there! :)

File: tests/unit/tests_sonata.cpp [hunk 4]
```diff
@@ -95,6 +95,22 @@ BOOST_AUTO_TEST_CASE( basicTestRotations )
...
+BOOST_AUTO_TEST_CASE( basicTestRotationsConverted )
+{
+    SonataFile file(SONATA_FILENAME_ALTERNATIVE);
+
+    Rotations neurons_rot = file.getRotations();
+
+    BOOST_CHECK_EQUAL(neurons_rot.shape()[0], 1000);
+    BOOST_CHECK( std::abs(neurons_rot[0][0] - 0.) < std::numeric_limits<double>::epsilon()*10 );
+    BOOST_CHECK( std::abs(neurons_rot[0][1] - -0.010005561303180392) < std::numeric_limits<double>::epsilon()*10 );
+    BOOST_CHECK( std::abs(neurons_rot[0][2] - 0) < std::numeric_limits<double>::epsilon()*10 );
```
[the remark this answers is not in the document]
Comment (matz-e) [reply tests/unit/tests_sonata.cpp]: Done

File: include/mvd/bits/sonata_misc.hpp (hunk 2 above)
Comment (matz-e) [reply include/mvd/bits/sonata_misc.hpp]: I think that will get pretty ugly. Not sure if it's worth it for code called once when loading.

File: include/mvd/bits/sonata_misc.hpp (hunk 3 above)
Comment (matz-e) [reply include/mvd/bits/sonata_misc.hpp]: Thanks, I split it.

**Pull request merged**